### ISSUE TITLE
[Bug]: NumPy version conflict with pandas and xarray

### ISSUE THREAD

**Issue by oscargus** [matplotlib/matplotlib#24884]
### Bug summary

When installing Matplotlib (e.g. in the tests)
```
  xarray 2022.12.0 requires numpy>=1.20, but you have numpy 1.19.3 which is incompatible.
  pandas 1.5.2 requires numpy>=1.20.3; python_version < "3.10", but you have numpy 1.19.3 which is incompatible.
```

### Code for reproduction

```python
python -m pip install -ve .
```

### Actual outcome

The warnings above

### Expected outcome

No warnings

### Matplotlib Version

main

### Installation

None

**Comment by rcomer** [comment]
Numpy 1.19 fell off the NEP29 schedule in June
https://numpy.org/neps/nep-0029-deprecation_policy.html#drop-schedule

**Comment by oscargus** [comment]
I do not fully get the mechanism here, but shouldn't `oldest-supported-numpy` then yield 1.20?

Example from one of the test runs:
```
  Collecting oldest-supported-numpy
    Downloading oldest_supported_numpy-2022.11.19-py3-none-any.whl (4.9 kB)
  Collecting pybind11>=2.6
    Downloading pybind11-2.10.2-py3-none-any.whl (222 kB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 222.1/222.1 kB 14.0 MB/s eta 0:00:00
  Collecting setuptools_scm>=7
    Using cached setuptools_scm-7.1.0-py3-none-any.whl (43 kB)
  Collecting numpy==1.19.3
    Using cached numpy-1.19.3-cp39-cp39-manylinux2010_x86_64.whl (14.9 MB)
```

https://github.com/matplotlib/matplotlib/blob/6091437be9776139d3672cde28a19cbe6c09dcd5/pyproject.toml#L5

Maybe we should at least update to require 1.20? (Or possibly 1.21 if we do not manage to get 3.7 out before Jan. 31...)

https://github.com/matplotlib/matplotlib/blob/6091437be9776139d3672cde28a19cbe6c09dcd5/setup.py#L325

**Comment by oscargus** [comment]
Ahh, `oldest-supported-numpy` is based on Python version, not NEP29...

**Comment by tacaswell** [comment]
> (Or possibly 1.21 if we do not manage to get 3.7 out before Jan. 31...)

We should bump to 1.20 in our requirements but drop off the cliff if we end up going past Jan 31 on getting 3.7 tagged.

**Comment by oscargus** [comment]
I updated locally and it, obviously, doesn't cause any issues, although I find it a bit strange that 1.19.3 is installed during installation. Not that people will notice it, but still.

(Also, the newly released scipy 1.10 requires 1.20, so a third warning is issued.)

**Comment by brobr** [comment]
Hi, when trying to upgrade matplotlib 3.6.2 to 3.7, I got an error because of missing dependency "oldest-supported-numpy" (with numpy 2.21 and python 3.9.16).

1) This dependency is not (yet?) mentioned in the list of [setup dependencies in the documentation](https://matplotlib.org/stable/devel/dependencies.html#setup-dependencies)
2) Installing that program did not help. Only after patching the pyproject.toml-file  to::

       diff -Naur matplotlib-3.7.0/pyproject.toml matplotlib-3.7.0b/pyproject.toml
       --- matplotlib-3.7.0/pyproject.toml	2023-02-13 17:59:24.000000000 +0000
       +++ matplotlib-3.7.0b/pyproject.toml	2023-02-17 16:27:48.113796144 +0000
       @@ -2,7 +2,7 @@
       build-backend = "setuptools.build_meta"
       requires = [
             "certifi>=2020.06.20",
       -    "oldest-supported-numpy",
       +    "numpy >= oldest-supported-numpy",
             "pybind11>=2.6",
             "setuptools_scm>=7",
             ]

compiling went ahead. This change might also address above issue of different numpy's expected.

hth

**Comment by tacaswell** [comment]
`oldest-supported-numpy` is not really a dependency, it is a trick to get the oldest version of numpy supported on the current Python (see https://pypi.org/project/oldest-supported-numpy/) without every project having to carry a wall of conditionals.

I suspect we missed updating this when we otherwise bumped the numpy version.

That anything was happy with `"numpy >= oldest-supported-numpy"` is...confusing to me and might be a bug in pip / build / wheel?

If this is in the slackware build process and you can patch, I would just patch that line out all together as the version of numpy you want to build with is the version of numpy that slackware ships (and will claim to be compatible with however that system works), not the oldest version of numpy that has wheels on pypi (given the `--no-isolation` I assume you have pre-installed all of the actual dependencies).

**Comment by brobr** [comment]
Thanks for the quick response;  ah well, python installs seem haemorrhaged with "tricks"; building from source is no longer fool-proof KISS, obviously. 

Yes, on Slackware all actual dependencies are installed beforehand when package management is done with system tools (and therefore a package can be installed on another computer with the same system). A pip-install into a virtual environment is a complete different route (but possible of course). Btw, Numpy, or Matplotlib etc. can be installed easily with tested build scripts from Slackbuilds.org; these programs are not part of the official Slackware distribution.

For such installs that bypass pip, the simplest patch would be to mention "numpy" after having an OK version of it pre-installed.

      -    "oldest-supported-numpy",
      +    "numpy",

That's why "anything was happy" I guess; the error that is otherwise thrown:

        ERROR Missing dependencies:
	oldest-supported-numpy
	numpy (==1.19.3)

gives a number below the numpy version installed.

Out of curiosity, can I ask in what way matplotlib has to face a "wall of conditionals"? Is giving a minimal version (say `"numpy >= 2.20",`  in the toml-file) not sufficient when matplotlib-coding depends on function-calls related to that version of numpy? (I hope to learn a bit to transfer it to my own work; my app just moved from `setup.cfg` to `pyproject.toml ` but I found it hard to see why this is so much better/easier/more pythonic for a user of python install tools.)

Anyway, thanks for matplotlib (everything still works on 3.7 ;-)

**Comment by rgommers** [comment]
If you want to drop older NumPy versions in the cadence that NEP 29 proposes, then you have to manage your version pins yourself, like SciPy does: https://github.com/scipy/scipy/blob/main/pyproject.toml#L21-L57. There is no `oldest-supported-numpy`-like metapackage for NEP 29. And that'd also be quite tricky to do, because the pins would have to be changed on certain dates, which then potentially breaks builds that were working before.

> Out of curiosity, can I ask in what way matplotlib has to face a "wall of conditionals"? Is giving a minimal version (say `"numpy >= 2.20",` in the toml-file) not sufficient when matplotlib-coding depends on function-calls related to that version of numpy?

It's not - see https://numpy.org/devdocs/dev/depending_on_numpy.html.

**Comment by brobr** [comment]
Thanks, those pointers are very helpful. The [NEP 29 info](https://numpy.org/neps/nep-0029-deprecation_policy.html), if that is what will be followed by Matplotlib, Numpy, Jupyter etc. makes some environmental changes more predictable.  

What makes it all unclear is that, when developing an app with numpy as *run-time* dependency, the info
>Most libraries that rely on NumPy will not need to set an upper version bound: NumPy is careful to preserve backward-compatibility.

suggests that an instruction `"numpy > oldest-supported-numpy"` would be okay while in the meantime questioning the '==' instruction for numpy as *building* requirement:
>Because for NumPy you have to care about ABI compatibility, you specify the version with == to the lowest supported version. For your other build dependencies you can probably be looser,..

This because, looking at that [scipy-paragraph](https://github.com/scipy/scipy/blob/main/pyproject.toml#L21-L57), preserving backward-compatibility would imply that numpy is not likely to throw a spanner in the wheel, but possibly the other conditionals numpy is linked to would (like  `python_version`, `platform_python_implementation`, `platform_machine` or `platform_system` ).

Apart from just a phrase, or trick,  "oldest-supported-numpy" seems a horse after the cart as it appears to deny the possibility that system-specific packages that back-port novel functions (say from python-3.11 to 3.9) would elevate the 'oldest-supported-numpy' number. In this case, as on Slackware happens, install stalls on this requirement for the wrong reason. Or, in other words, because it does not account for a version control system that deviates from that of ubuntu, macos, windows, conda etc., it is not uniformly applicable and therefore too restrictive. This seems contrary to the spirit of Python, or Matplotlib.

**Comment by rgommers** [comment]
> suggests that an instruction `"numpy > oldest-supported-numpy"` would be okay

That's simply not a valid version specifier. If all you need is a runtime dependency, use `numpy>=1.22.3` or some such number. And if you need a dynamic runtime version which depends on the build-time version,  then you unfortunately have to do build-time string munging such that you produce metadata (like `install_requires`) with a concrete version specifier like `1.22.3`. It must be a [PEP 508](https://peps.python.org/pep-0508/) compliant specifier, which `> oldest-supported-numpy` is not.

> Apart from just a phrase, or trick, "oldest-supported-numpy" seems a horse after the cart as it appears to deny the possibility that system-specific packages that back-port novel functions (say from python-3.11 to 3.9) would elevate the 'oldest-supported-numpy' number. ...

I'm sorry, I am not really sure what you are trying to express. However I interpret all that though, it seems wrong. There's no such thing as backporting new functions from Python 3.11 to 3.9.

**Comment by brobr** [comment]
Thanks again for the feedback; Resorting to concrete version specifiers makes sense for patching the pyproject.toml file.

 > There's no such thing as backporting new functions from Python 3.11 to 3.9.

My description could be more precise; functionality lacking in the Python library shipped with a system like Slackware can, when available, be compensated with 'backports'. For example, [importlib-resources]( https://slackbuilds.org/repository/15.0/python/importlib-resources/), which turned up  on my slackware box as a new dependency to upgrade from 3.6.2 to matplotib 3.7.0.

The major issue is usability of tools to get python-packages installed. And one that fails the bill seems  `oldest-supported-numpy` ; it is not an importable python module and it doesn't allow for leeway that numpy provides with it's backward compatibility.

**Comment by rgommers** [comment]
> The major issue is usability of tools to get python-packages installed.

@brobr this does not seem like the right place for a general complaint about usability of Python packaging tooling. Your question has been answered by @tacaswell already. I think we all agree that it's a bit messy and wish packaging was easier, but given where it's at now, Matplotlib must have a build-time dependency like `oldest-supported-numpy`. 

> For such installs that bypass pip, the simplest patch would be to mention "numpy" after having an OK version of it pre-installed.

This is a bigger misunderstanding by the way. Your distro packaging tooling must use `pip install . --no-build-isolation` (or similar), and manage all dependencies _outside of `pyproject.toml`_. You cannot expect to reuse pypi-specific dependency specifications as-is for Linux distro packaging. Most distros have tooling for this like `pypi2rpm`.

**Comment by timhoffm** [comment]
AFAICS there‘s nothing to do from our side.
